### PR TITLE
linters: add Godox linter support

### PR DESCRIPTION
**Pull request by matoous**
Godox is linter for TODOs and FIXMEs left in the code.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -192,7 +192,7 @@ gosimple: Linter for Go source code that specializes in simplifying a code [fast
 govet (vet, vetshadow): Vet examines Go source code and reports suspicious constructs, such as Printf calls whose arguments do not align with the format string [fast: false, auto-fix: false]
 ineffassign: Detects when assignments to existing variables are not used [fast: true, auto-fix: false]
 staticcheck: Staticcheck is a go vet on steroids, applying a ton of static analysis checks [fast: false, auto-fix: false]
-structcheck: Finds an unused struct fields [fast: true, auto-fix: false]
```
Comment (Sean-Der) [inline README.md]: I would include this change in another PR

File: README.md [hunk 2]
```diff
@@ -204,6 +204,7 @@ and the following linters are disabled by default:
 $ golangci-lint help linters
 ...
 Disabled by default linters:
+bodyclose: checks whether HTTP response body is closed successfully [fast: false, auto-fix: false]
```
Comment (Sean-Der) [inline README.md]: I would include this change in another PR

File: README.md [hunk 3]
```diff
@@ -458,7 +460,7 @@ Usage:
   golangci-lint run [flags]
 
 Flags:
-      --out-format string           Format of output: colored-line-number|line-number|json|tab|checkstyle|code-climate (default "colored-line-number")
+      --out-format string           Format of output: colored-line-number|line-number|json|tab|checkstyle|code-climate|junit-xml (default "colored-line-number")
```
Comment (Sean-Der) [inline README.md]: I would include this change in another PR

File: README.md [hunk 4]
```diff
@@ -525,6 +527,7 @@ Global Flags:
       --cpu-profile-path string   Path to CPU profile output file
       --mem-profile-path string   Path to memory profile output file
   -v, --verbose                   verbose output
+      --version                   Print version
```
Comment (Sean-Der) [inline README.md]: I would include this change in another PR

File: pkg/config/config.go [hunk 5]
```diff
@@ -293,7 +303,8 @@ type Issues struct {
 	NeedFix bool `mapstructure:"fix"`
 }
 
-type Config struct { //nolint:maligned
+type Config struct {
+	// nolint:maligned
```
Comment (Sean-Der) [inline pkg/config/config.go]: I would include this in another PR

File: README.md (hunk 1 above)
Comment (matoous) [reply README.md]: Fixed!

File: README.md (hunk 2 above)
Comment (matoous) [reply README.md]: Fixed.

File: README.md (hunk 3 above)
Comment (matoous) [reply README.md]: FIxed.

File: README.md (hunk 4 above)
Comment (matoous) [reply README.md]: Removed.

File: pkg/config/config.go (hunk 5 above)
Comment (matoous) [reply pkg/config/config.go]: Done.

**Comment by matoous** [comment]
Hey, I fixed the issues, can you review this once again @Sean-Der ? Thansk!

**Comment by jirfag** [comment]
Hi, thank you!
Didn't you forget to include `golinters/godox.go` file?

**Comment by jirfag** [comment]
also, it looks like that it should be rebased on the fresh master branch

**Comment by matoous** [comment]
@jirfag fixed everything I hope

File: pkg/config/config.go [hunk 6]
```diff
@@ -211,6 +212,11 @@ type PreallocSettings struct {
 	ForLoops   bool `mapstructure:"for-loops"`
 }
 
+type GodoxSettings struct {
+	IncludeTests bool `mapstructure:"include-tests"`
```
Comment (jirfag) [inline pkg/config/config.go]: looks like it's unused. And there is no need for such a setting because we can configure whether to run any specific linter on test files by `exclude-rules`

File: pkg/golinters/godox.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"context"
+	"go/token"
+	"strings"
+
+	"github.com/golangci/golangci-lint/pkg/lint/linter"
+	"github.com/golangci/golangci-lint/pkg/result"
+
+	"github.com/matoous/godox"
```
Comment (jirfag) [inline pkg/golinters/godox.go]: I see `https://github.com/matoous/godox` is a fork. Can you please explain in your fork's README reasoning for forking?

File: .golangci.example.yml [hunk 8]
```diff
@@ -186,6 +186,10 @@ linters-settings:
         paramsOnly: true
       rangeValCopy:
         sizeThreshold: 32
+  godox:
+    keywords:
```
Comment (jirfag) [inline .golangci.example.yml]: please, leave a comment here that default keywords are `TODO`, `BUG`, `FIXME` and here we can specify which keywords we need and they will override default keywords.

File: .golangci.example.yml [hunk 9]
```diff
@@ -186,6 +186,10 @@ linters-settings:
...
+      - FIXME
+      - TODO
```
Comment (jirfag) [inline .golangci.example.yml]: please use some practical example - it would be practical if it would be different from the default keywords

File: .golangci.yml [hunk 10]
```diff
@@ -41,6 +41,10 @@ linters-settings:
     disabled-checks:
       - wrapperFunc
       - dupImport # https://github.com/go-critic/go-critic/issues/845
+  godox:
```
Comment (jirfag) [inline .golangci.yml]: please, remove it: these keywords are already enabled by default

File: test/testdata/godox.go [hunk 11]
```diff
@@ -0,0 +1,7 @@
+//args: -Egodox
+//config: linters-settings.godox.keywords=FIXME,TODO
+package testdata
+
+func todoLeftInCode() {
+	// TODO implement me // ERROR godox.go:6: Line contains FIXME/TODO: "TODO implement me"
```
Comment (jirfag) [inline test/testdata/godox.go]: please, add more tests:
```
//TODO no space
// TODO(author): 123
//TODO(author): 123
//TODO(author) 456
// TODO: qwerty
// todo 789
```

File: go.mod [hunk 12]
```diff
@@ -56,7 +57,7 @@ require (
 	github.com/valyala/quicktemplate v1.1.1
 	golang.org/x/crypto v0.0.0-20190313024323-a1f597ede03a // indirect
 	golang.org/x/sys v0.0.0-20190312061237-fead79001313 // indirect
-	golang.org/x/tools v0.0.0-20190909030654-5b82db07426d
+	golang.org/x/tools v0.0.0-20190910044552-dd2b5c81c578
```
Comment (jirfag) [inline go.mod]: can you do it in a separate pull request please? it will be easier to review and merge

Comment (matoous) [reply go.mod]: Yes, removed.

File: test/testdata/godox.go (hunk 11 above)
Comment (matoous) [reply test/testdata/godox.go]: Added. My reasoning was that tests for all these cases are in the repository with `godox` and as such are not needed here.

File: .golangci.yml (hunk 10 above)
Comment (matoous) [reply .golangci.yml]: Removed.

File: .golangci.example.yml (hunk 9 above)
Comment (matoous) [reply .golangci.example.yml]: Added examples and comments.

File: .golangci.example.yml (hunk 8 above)
Comment (matoous) [reply .golangci.example.yml]: Added.

File: pkg/config/config.go (hunk 6 above)
Comment (matoous) [reply pkg/config/config.go]: Removed.

File: pkg/config/config.go [hunk 13]
```diff
@@ -230,6 +235,9 @@ var defaultLintersSettings = LintersSettings{
 	Gocritic: GocriticSettings{
 		SettingsPerCheck: map[string]GocriticCheckSettings{},
 	},
+	Godox: GodoxSettings{
+		Keywords: []string{"FIXME", "TODO"},
```
Comment (jirfag) [inline pkg/config/config.go]: it looks like default values are duplicated in two places: here and in [godox repo](https://github.com/matoous/godox/blob/master/godox.go#L14). And they differ :) Maybe remove defaults here and leave only defaults in godox?

File: vendor/modules.txt [hunk 14]
```diff
@@ -198,7 +200,7 @@ golang.org/x/sys/windows
 golang.org/x/text/transform
 golang.org/x/text/unicode/norm
 golang.org/x/text/width
-# golang.org/x/tools v0.0.0-20190909030654-5b82db07426d => github.com/golangci/tools v0.0.0-20190910062050-3540c026601b
+# golang.org/x/tools v0.0.0-20190910044552-dd2b5c81c578 => github.com/golangci/tools v0.0.0-20190910062050-3540c026601b
```
Comment (jirfag) [inline vendor/modules.txt]: run `make vendor` to remove it please

File: pkg/golinters/godox.go (hunk 7 above)
Comment (matoous) [reply pkg/golinters/godox.go]: Updated! Good point.

File: go.mod (hunk 12 above)
Comment (matoous) [reply go.mod]: Fixed.

File: pkg/config/config.go (hunk 13 above)
Comment (matoous) [reply pkg/config/config.go]: Removed.

File: vendor/modules.txt (hunk 14 above)
Comment (matoous) [reply vendor/modules.txt]: `make vendor` didn't work as it modified the `go.mod` and this file to what it was. Had to undo it manually. Fixed now.

**Comment by Sean-Der** [comment]
Great work @matoous 🎉 the second this comes out in a release I will start using it.

Good to see you back @jirfag, you are really doing a crazy amount of work :)
